src/_pytest/python.py

```diff
@@ -929,6 +929,130 @@ def hasnew(obj: object) -> bool:
     return False
 
 
+@final
+@attr.s(frozen=True, auto_attribs=True, slots=True)
+class IdMaker:
+    argnames: Sequence[str]
+    parametersets: Sequence[ParameterSet]
+    idfn: Optional[Callable[[Any], Optional[object]]]
+    ids: Optional[Sequence[Union[None, str]]]
+    config: Optional[Config]
+    nodeid: Optional[str]
+
+    def make_unique_parameterset_ids(self) -> List[str]:
+        """Make a string for each ParameterSet to identify test invocations in pytest logs.
+
+        Format is <prm_1_token>-...-<prm_n_token>[counter],
+        where prmset_x_token is
+        - the parameter id, derived from value or user provided id, if applicable
```

Comment:
Would split this to two:

- user-provided id, if given
- id derived from the value, for certain types